state/src/behavior.ts

```diff
@@ -0,0 +1,181 @@
+
+  static define<Spec, Value = Spec>({combine, behavior = noBehavior, default: default_ = noDefault}: {
+    combine: (specs: A<Spec>) => Value,
+    behavior?: (value: any) => A<BehaviorSpec>,
+    default?: Spec
+  }) {
+    return new Behavior<Spec, Value>(combine, behavior, default_)
+  }
+
+  static defineSet<Spec>({behavior = noBehavior, default: default_ = noDefault}: {
+    behavior?: (spec: Spec) => A<BehaviorSpec>,
+    default?: Spec
+  } = {}) {
+    return new SetBehavior<Spec>(null, behavior, default_)
+  }
+
+  use(spec: Spec = this.default_, priority: Priority = noPriority): BehaviorSpec {
+    if (spec == noDefault) throw new RangeError("This behavior has no default spec")
+    return new BehaviorSpec(this, spec, priority)
+  }
+
+  get(state: EditorState): Value | undefined {
+    return state.config.behavior.get(this)
+  }
+
+  static stateField: SetBehavior<StateField<any>>
+
+  static multipleSelections = Behavior.define<boolean, boolean>({
+    combine: values => values.indexOf(true) > -1,
+    default: true
+  })
+
+  // FIXME move to view?
+  static viewPlugin: SetBehavior<(view: any) => any>
+
+  static indentation: SetBehavior<(state: EditorState, pos: number) => number>
+
+  // @internal
+  hasSubBehavior(behavior: Behavior<any>): boolean {
+    for (let sub of this.knownSub)
+      if (sub == behavior || sub.hasSubBehavior(behavior)) return true
+    return false
+  }
+
+  // @internal
+  getBehavior(input: any, priority: Priority): A<BehaviorSpec> {
+    let sub = this.behavior(input)
+    for (let b of sub)
+      if (this.knownSub.indexOf(b.type) < 0)
+        this.knownSub.push(b.type)
+    return sub.map(b => b.fillPriority(priority))
+  }
+
+  // Utility function for combining behaviors to fill in a config
+  // object from an array of provided configs. Will, by default, error
+  // when a field gets two values that aren't ===-equal, but you can
+  // provide combine functions per field to do something else.
+  static combineConfigs<Config>(configs: A<Config>,
+                                combine: {[key: string]: (first: any, second: any) => any} = {},
+                                defaults?: Config): Config {
+    let result: any = {}
+    for (let config of configs) for (let key of Object.keys(config)) {
+      let value = (config as any)[key], current = result[key]
+      if (current === undefined) result[key] = value
+      else if (current === value || value === undefined) {} // No conflict
+      else if (Object.hasOwnProperty.call(combine, key)) result[key] = combine[key](current, value)
+      else throw new Error("Config merge conflict for field " + key)
+    }
+    if (defaults) for (let key in defaults)
+      if (result[key] === undefined) result[key] = (defaults as any)[key]
+    return result
+  }
+}
+
+export class SetBehavior<Spec> extends Behavior<Spec, A<Spec>> {
+  get(state: EditorState): A<Spec> {
+    return state.config.behavior.get(this) || none
+  }
+}
```

Comment:
Maybe consider removing this subclass and move the check to the callers (currently commands and editorview). Or instead generalize a default value concept for non-set behaviors, too?

Reply:
I also consider having this class a bit ugly, but the alternatives seem more problematic. For non-set behaviors, being present or not are distinct conditions, and thus we'd require some kind of special null value there, not a default.

Maybe make `get` error when called on a non-set behavior that isn't present, and provide a different accessor to check for availability?